substreams/ethereum-nabla/src/modules/map_components.rs

```diff
@@ -0,0 +1,195 @@
+use crate::abi;
+use anyhow::Result;
+use itertools::Itertools;
+
+use substreams::hex;
+use substreams_ethereum::pb::eth::v2::{Log, TransactionTrace};
+use substreams_ethereum::{pb::eth, Event};
+use substreams_helper::hex::Hexable;
+use tycho_substreams::prelude::{
+    Attribute, BlockTransactionProtocolComponents, ChangeType, FinancialType, ImplementationType,
+    ProtocolComponent, ProtocolType, TransactionProtocolComponents,
+};
+
+struct Portal {
+    contract_address: [u8; 20],
+    tx_hash: [u8; 32],
+}
+
+const NABLA_PORTAL: Portal = Portal {
+    contract_address: hex!("cB94Eee869a2041F3B44da423F78134aFb6b676B"),
+    tx_hash: hex!("7e9c5f39e41080aa7ab891ddd8669efc7191bae5906f74c2d5c7e7c12e219046"),
+};
+
+#[derive(Debug, Clone)]
+pub enum ComponentType {
+    Portal,
+    Router,
+    SwapPool,
+}
+
+impl Into<String> for ComponentType {
+    fn into(self) -> String {
+        match self {
+            ComponentType::Portal => "portal".into(),
+            ComponentType::Router => "router".into(),
+            ComponentType::SwapPool => "swap_pool".into(),
+        }
+    }
+}
+
+impl From<String> for ComponentType {
+    fn from(s: String) -> Self {
+        match s.as_str() {
+            "portal" => ComponentType::Portal,
+            "router" => ComponentType::Router,
+            "swap_pool" => ComponentType::SwapPool,
+            _ => panic!("Unknown protocol type: '{}'", s),
+        }
+    }
+}
```

Comment:
these need to be stored as `Attribute`s of the `ProtocolComponent`, which are immutable. Since these are `String`, and not `Enum` type, this offers a clean way to translate between them (panic should never occur based on the logic, and this prevents the use of hardcoded `String`s in different modules, which is nasty)